File: commons/base/src/main/java/com/hb/base/rx/RxSearch.java
package com.hb.base.rx;

import android.text.Editable;
import android.text.TextWatcher;
import android.widget.EditText;

import androidx.appcompat.widget.SearchView;

import io.reactivex.Observable;
import io.reactivex.subjects.PublishSubject;

/**
 * Created by buihai on 1/12/18.
 */

public class RxSearch {

  public static Observable<String> fromView(SearchView view) {

    final PublishSubject<String> subject = PublishSubject.create();

    view.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
      @Override
      public boolean onQueryTextSubmit(String query) {
//                subject.onComplete();
        return true;
      }

      @Override
      public boolean onQueryTextChange(String newText) {
        subject.onNext(newText);
        return true;
      }
    });

    return subject;
  }

  private static class MyTextWatcher implements TextWatcher {

    final PublishSubject<String> subject;

    MyTextWatcher(PublishSubject<String> subject) {
      this.subject = subject;
    }


    @Override
    public void beforeTextChanged(CharSequence s, int start, int count, int after) {

    }

    @Override
    public void onTextChanged(CharSequence s, int start, int before, int count) {
      subject.onNext(s.toString());
    }

    @Override
    public void afterTextChanged(Editable s) {

    }
  }

  private static TextWatcher watcher = null;


  public static Observable<String> fromView(EditText view) {

    final PublishSubject<String> subject = PublishSubject.create();

    if (watcher == null) {
      watcher = new MyTextWatcher(subject);
      view.addTextChangedListener(watcher);
    }

    return subject;
  }
}

File: commons/base/src/main/java/com/hb/base/rx/RxSearch.kt
package com.hb.base.rx

import android.text.Editable
import android.text.TextWatcher
import android.widget.EditText
import androidx.appcompat.widget.SearchView
import io.reactivex.Observable
import io.reactivex.subjects.PublishSubject

/**
 * Created by buihai on 1/12/18.
 */
object RxSearch {
  fun fromView(view: SearchView): Observable<String> {
    val subject = PublishSubject.create<String>()
    view.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
      override fun onQueryTextSubmit(query: String): Boolean {
//                subject.onComplete();
        return true
      }

      override fun onQueryTextChange(newText: String): Boolean {
        subject.onNext(newText)
        return true
      }
    })
    return subject
  }

  private var watcher: TextWatcher? = null
  fun fromView(view: EditText): Observable<String> {
    val subject = PublishSubject.create<String>()
    if (watcher == null) {
      watcher = MyTextWatcher(subject)
      view.addTextChangedListener(watcher)
    }
    return subject
  }

  private class MyTextWatcher internal constructor(val subject: PublishSubject<String>) :
    TextWatcher {
    override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
    override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
      subject.onNext(s.toString())
    }

    override fun afterTextChanged(s: Editable) {}
  }
}
